pkg/controller/kubedirectorcluster/members.go

```diff
@@ -424,6 +424,34 @@ func handleReadyMembers(
 
 			m.StateDetail.LastConfigDataGeneration = cr.Status.SpecGenerationToProcess
 		}(member)
+
+		updateRoleUpgradeStatus := func() {
```

Comment:
I'm dubious that this is the right place to run this code, for a couple of reasons.

First thing: You currently have it inside the loop that is iterating over every ready member in the role. So if there are for example three ready members then updateRoleUpgradeStatus will be run three times in this loop, which seems overkill since each call of updateRoleUpgradeStatus iterates over all the role members.

So maybe one thing is that you can/should just pop it outside of this ready-members loop. However I think that still leaves the question of whether we should be doing this role status management inside handleReadyMembers at all. (It doesn't seem to be related to the "mission" of updating ready-status members with new configmeta.) You're setting the final pod upgrade status as part of handleCreatingMembers... maybe setting the role upgrade status should go there too?

Reply:
Got it, I moved the logic to `syncRoles` method.